server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -748,21 +749,23 @@ private static void validate(final String repositoryName, final String snapshotN
         }
     }
 
-    private static ShardGenerations buildGenerations(SnapshotsInProgress.Entry snapshot, Metadata metadata) {
+    private static UpdatedShardGenerations buildGenerations(SnapshotsInProgress.Entry snapshot, Metadata metadata) {
         ShardGenerations.Builder builder = ShardGenerations.builder();
+        ShardGenerations.Builder deletedBuilder = ShardGenerations.builder();
         if (snapshot.isClone()) {
             snapshot.shardSnapshotStatusByRepoShardId().forEach((key, value) -> builder.put(key.index(), key.shardId(), value));
         } else {
             snapshot.shardSnapshotStatusByRepoShardId().forEach((key, value) -> {
                 final Index index = snapshot.indexByName(key.indexName());
                 if (metadata.findIndex(index).isEmpty()) {
                     assert snapshot.partial() : "Index [" + index + "] was deleted during a snapshot but snapshot was not partial.";
+                    deletedBuilder.put(key.index(), key.shardId(), value);
                     return;
                 }
                 builder.put(key.index(), key.shardId(), value);
             });
         }
-        return builder.build();
+        return new UpdatedShardGenerations(builder.build(), deletedBuilder.build());
```

Comment:
`deletedBuilder.build()` is mostly going to be `ShardGenerations.EMPTY`, can we special-case the empty collection in `ShardGenerations.Builder` to skip a bunch of unnecessary allocations?

Reply:
Makes sense. Special cased in both `Builder#build` and initialization of the variable here 81f88baa32d7ce775086efaf6b6722e370d2f6e5